### DISCUSSION TITLE
Any plans to release a pre-compiled binary?

### DISCUSSION THREAD

**Discussion by 8465231** [pete4abw/lrzip-next#19] [Q&A] [answered]
You were helping me in the main LRzip repo.

I found this when looking at your account and it looks real interesting.

Any plans to release a pre-compiled version (or even better a dockertized version combined with lrzip-fee)? 

I have no idea where to start for compiling, have not done it in years and then was a clear case of failing a few dozen times and not having a clue why it finally worked, then deciding it was just not worth the hassle lol.

Would this version net me any real world gains in compression / speed for my 2.5TB file using lmza compression?

lrzip-fe also looks FANTASTIC! While I will happily run scripts and such from CLI, give me a GUI for anything that I don't do repeatedly any day of the week.

Thanks for the help BTW. I love learning.

**Comment by pete4abw** [comment]
No no plans for pre-compiled. `configure, make, make install` are not beyond most. Docker? I don't see the need yet. Back to your original issue, I strongly recommend using 
`tar -I 'lrzip [options] -cf file.tar.lrz files...` 
or pipe tar to lrzip, 
`tar -cf - files... | lrzip [options]  -o file.tar.lrz`

Why you would waste time tar'ring a huge file, basically doubling your disk use, is unclear. When you use `tar` and lrz you can actually select which files/directories to extract. If you did the tarfile lrzip file route, you would have to extract your huge 2.5TB file and then selectively extract.

I recommend reading through some of the READMEs, Benchmarks, the [Wiki](https://github.com/pete4abw/lrzip-next/wiki) here and [Discussions](https://github.com/pete4abw/lrzip-next/discussions). Then do small scale tests to see about benefits.

This [README](https://github.com/pete4abw/lrzip-next/blob/master/doc/README.NEW.BENCHMARK.ALGO.md) in the doc directory is helpful in selecting the best compression.

Good luck and thank you for your interest.

**Comment by 8465231** [comment]
You greatly overestimate the abilities of linux noobs. lol

First issue, don't have git installed or know how to use it. I figured that out after some googling though. 

Hint, this is basically the issues I always have with linux, no matter what I want to do I always seem to be missing the right package or dependency and can never seem to find the right names to install it. Leading to an endless google loop just to use one program.

With docker I run 1 command that has everything all setup and ready to go in seconds regardless of skill level. Docker is the single reason I have even ventured back into linux (well, that and unraid). Off topic but I know it is easy to forget what it is like not even knowing what to search for to find the answer you want.

Just to give you an idea what a noob goes through for a "simple" command like this here is what I spent the last 45 minutes doing:

Second issue:

/github/lrzip-next# ./autogen.sh
Running autoreconf -if...
./autogen.sh: 8: autoreconf: not found

back to google...

Fixed that by installing autoreconf but then ran into this problem:

Running autoreconf -if...
./util/gitdesc.sh: 50: Bad substitution
./util/gitdesc.sh: 50: Bad substitution
./util/gitdesc.sh: 50: Bad substitution
........
./util/gitdesc.sh: 50: Bad substitution
./util/gitdesc.sh: 50: Bad substitution
configure.ac:20: error: possibly undefined macro: AC_PROG_LIBTOOL
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
autoreconf: /usr/bin/autoconf failed with exit status: 1

Google is not being helpful with this one, m4_pattern_allow is not helping. It looks like AC_PROG_LIBTOOL is a package but apt install just says:

E: Unable to locate package AC_PROG_LIBTOOL

After a lot of searching I found an indirect answer that said I need to install libtool, this at least made a difference but still spit out a ton of errors.

Thankfully the script spit out this with the next error:

configure: error: Could not find zlib library - please install zlib-dev

Sadly this did not work and gave this error: E: Unable to locate package zlib-dev

So back to google I went, more searching with no real answers finally led to finding the correct name for the package was zlib1g-dev

Next error is configure: error: Could not find bz2 library - please install libbz2-dev

Thankfully this one was correct but still a ton of errors and a new error:

WARNING:
The doxygen program was not found in your execute path.
You may have doxygen installed somewhere not covered by your path.

If this is the case make sure you have the packages installed, AND
that the doxygen program is in your execute path (see your
shell manual page on setting the $PATH environment variable), OR
alternatively, specify the program to use with --with-doxygen.
configure: WARNING: no doxygen detected. Documentation will not be built
checking that generated files are newer than configure... done

Back to google I go. Install doxygen and try again.

Next error, this one much harder to figure out, I am sure something is missing but no hints on what it could be.

config.status: error: Something went wrong bootstrapping makefile fragments
    for automatic dependency tracking.  Try re-running configure with the
    '--disable-dependency-tracking' option to at least be able to build
    the package (albeit without support for automatic dependency tracking).
See `config.log' for more details

config.log is useless to me.

I am also not finding much help on google. Others had the issue but it seemed to be the wrong version of a package not a missing package. 

I start installing every package I see mentioned but it is not helping. Still gets the same error.

After 45 minutes, I am going to have to take a break. lol

FYI, this is not aimed at you, more just a rant about trying to do things on linux in a general sense. This is my linux life pretty much, I spend more time on google then doing anything on linux until I discovered docker and the skies opened, I could actually run programs without wanting to throw the computer at the wall! lol

I am going to read those links you posted now, thanks for those, I was not even aware that github had a discussion section, this should of been posted there I am sure.

Sorry for posting this in the wrong place. I am used to discussion forums.

**Comment by pete4abw** [comment]
I feel your pain.  But As the Readme says, you must clone the repo in order to compile. Downloading a tarball will not build. There's a reason for this since the version is dependant on specific commits and if something breaks it's helpful to know the precise version built.


 If you're going to build your own package, you need most of not all development libraries. Normally this is an installation option. Since lrzip-next is not offered as a package by distros, evaluators must build and install by themselves.

My advice to you is install the development packages for Ubuntu. Git will be included.

Discussions are on the top level menu. Better place for this. I'm moving this over...

Regards

**Comment by 8465231** [comment]
Yeah, I did install git and used the command in the readme to clone it. That was not that hard to figure out since the command was there.

If there is an option that does not include building my own package I am all ears, it looked like this was the only option?

What are the development libraries? I have installed everything I have seen mentioned in google at this point and still getting a long list of this error:

./util/gitdesc.sh: 50: ./util/gitdesc.sh: Bad substitution

It did finally complete the ./autogen.sh command though and I tried make [-j8] but it gave this error:

make: *** No rule to make target '[-j8]'.  Stop.

I then tried make install and it did a bunch of stuff but still errored out with this:

lrzip.c:135:42: error: expected expression before ',' token
  135 |   'L', 'R', 'Z', 'I', LRZIP_MAJOR_VERSION, LRZIP_MINOR_VERSION
      |                                          ^
lrzip.c: In function 'get_magic':
lrzip.c:202:51: error: expected expression before '||' token
  202 |  if (control->major_version > LRZIP_MAJOR_VERSION ||
      |                                                   ^~

Oh the joys of not having a clue what you are doing. lol.

**Reply by pete4abw** [reply]
The [ -j8] means optionally use the `-j8` option not literally use the brackets. I think for the time being, use the standard lrzip which Ubuntu provides. The wiki will help you though.

**Reply by 8465231** [reply]
When you don't understand what you are doing, you just do whatever you are told. lol

Is there a list of development dependencies that I can try installing? I have installed everything that I can find but still get a bunch of these errors

./util/gitdesc.sh: 50: ./util/gitdesc.sh: Bad substitution

And make fails as well naturally.

Gonna take a break and read those links you posted, it would be nice to try this new version while testing things anyways but maybe it is not meant to be. Annoyed that I can't get it working after almost 2 hours at this point.

**Comment by 8465231** [comment]
on another topic, I was not aware it was possible to extract a specific file from a lrzip tar file as you mentioned, how do you do that?

**Reply by pete4abw** [reply]
If you pipe tar to lrzip-next, the lrz file will encapsulate the tar file. Then you extract the tar reversing three commands. See the lrzip-next wiki. There's an article on it.

**Comment by 8465231** [comment]
Yeah, I have not gotten to the wiki yet, still trying to get this to compile, can't get past this latest error.

Looks like I might be stuck with the regular lrzip sadly.

**Comment by 8465231** [comment]
I just can't figure out this error when running autogen.sh: 

./util/gitdesc.sh: 50: ./util/gitdesc.sh: Bad substitution

Google keeps pointing towards an issue with bash vs sh but I have tried both and all the suggestions I can find on the subject but nothing changes.

make keeps failing as well, seems that it can't find 

rzip.c:40:10: fatal error: lzo/lzoconf.h: No such file or directory
   40 | #include <lzo/lzoconf.h>

Then it has this error:

rzip.c:135:42: error: expected expression before ',' token
  135 |   'L', 'R', 'Z', 'I', LRZIP_MAJOR_VERSION, LRZIP_MINOR_VERSION

I can't find anything useful on locating the lzo.conf. I see it is in the git clone under src/lib/minilzo/lzoconf.h

Does this dir need to be renamed to lzo?

**Reply by pete4abw** [reply]
seems like `configure` is not completing due to errors, missing libraries, etc. The rzip error you see is because the file config.h does not exist. You need to pay attention to your error messages.

Anyway, you make a decent point about creating packages. Since this is a development version and experimental at that, I am not inclined at present.

It's hard to jump into development without experience, and in a new OS besides where everything is different. You're missing many tools and libraries that are normally installed when you select full or Development install options. I don't use Ubuntu or Debian, so I cannot help much.

No lco.conf is part of the lzo library which is normally installed in most distros. Why it is not installed in yours is not clear. It's likely that you did not install the development libraries which will include header files, tools, etc.

See your Ubuntu documentation for those install instructions.

My suggestion to you is to use lrzip and tar as explained in the wiki, don't try and tar 2.5TB and compress it - that's dumb. Rather employ a strategy to backup pieces at a time. And don't use -U, -z, -9 options. In fact, as the benchmarks show, levels 4 and 6 give you the best bang for the buck in terms of compression and time.

**Reply by 8465231** [reply]
Ok, I will just go back to the normal lrzip. I was really looking forward to the lrzip-fe. A gui version would be such a breath of fresh air.

I am not really worried about compression time as long as it is reasonable (aka, less then a week), size is more important. That said I think I will stop using the -L 9 option.

-U I am simply testing at this point. In smaller tests I saw a 10-20% improvement in size, I figured that could be even greater with a large file like this that should have a lot of redundant information.

My plan was to try it again without the -U option to see the real world effects.

I thought about breaking it up into smaller chunks but there is no easy way to break it up without getting really unorganized. 

Thanks for your help!

**Reply by pete4abw** [reply]
You can use lrzip-fe. Just that some options won't apply, like filtering. It's a great way to explore options without actually having to run the app. If you do want to run it, you can either copy the command line or run lrzip from inside a window.

**Comment by 8465231** [comment]
afraid to even say this but I can't get the lrzip-fe to run either.

Tried it on 2 separate systems, one is a fresh install of ubuntu 20.04 LTS which I get this error on:

./lrzip-fe.sh: line 7: syntax error near unexpected token `newline'
./lrzip-fe.sh: line 7: `<!DOCTYPE html>'

the other is a ubuntu docker that gives this error:

./lrzip-fe.sh 
./lrzip-fe.sh: 41: dialog: not found
./lrzip-fe.sh: 48: [: -lt: unexpected operator
./lrzip-fe.sh: 51: let: not found
./lrzip-fe.sh: 53: [: -lt: unexpected operator
./lrzip-fe.sh: 56: let: not found
./lrzip-fe.sh: 80: let: not found
./lrzip-fe.sh: 82: let: not found
./lrzip-fe.sh: 83: [: -lt: unexpected operator
./lrzip-fe.sh: 87: dialog: not found
./lrzip-fe.sh: 93: exit: Illegal number: -1

Oh well, if a docker is ever made for these I would be all over it. I have always enjoyed compressing things, not sure why lol.

I will never understand why docker gets the hate it does from some. I understand it is not perfect but boy is it simple. A great way for guys like me to get my feet in the door so I can learn at my own rate. 

Linux is an all or nothing OS without it, you either know everything or you flounder not even knowing what you don't know. For example I have spent the last half hour trying to find a list of all dev tools so I can install them but outside of a few incomplete lists, have yet to find anything that works. I don't even know what to search for. /rant

Thanks for your help, I have taken enough of your time at this point I think.

**Reply by pete4abw** [reply]
How are you even downloading it? if `dialog` is not found you can't run lrzip-fe. dialog is the curses program that manages the screen. The errors displayed are indicative of your not having a bash shell. I've no idea what you have going on. I honestly think you need to start with basics and learn linux, then learn bash, and certainly master installation of linux. You've got a partial installation which won't do much other than run programs already packaged. If you need to develop or compile, you will not be able to.

**Reply by 8465231** [reply]
I just ran it from the cloned git folder. there were no instructions that I could see on a proper method for running it.

Yeah, I figured that second one was due to something like that, I just tried it in an ubuntu docker to see if it made a difference. 

When installing ubuntu I just picked the LTS version and basically left everything on default. 

Honestly I have no desire to compile or develope things as every time I try I have issues like this, it is never fun lol. Sadly it seems to be required to do anything outside of the basics on linux so I am being tossed into the deep end.

Do you know of a distro or docker that has a complete development enviroment with everything I could need already setup? I would be all over that, so far my searching has not turned up anything.

Just "this distro is great for devs, you can install all the dev tools from ppa's". That doesn't help when you don't have a clue what those tools are you need to install.

**Reply by pete4abw** [reply]
I've been using one distro for over 20 years. It's a very do-it-yourself distro and not appropriate for you. Dev tools are normally installed as a bunch, not one by one. Installation normally asks you what type of install you want. You likely got a desktop installation that did not include development libraries. Sorry I can't help more on this.

**Comment by 8465231** [comment]
Ok, thats fine, thank you for your time and the help.

So when looking for a development distro, those are the keywords I should use, "development" along with "distro" or "tools". I am just not getting the results I would expect when searching with these.

**Reply by pete4abw** [reply]
All distros have this. You have to install properly. A full installation normally does this. RTFM.

**Comment by pete4abw** [comment, marked as answer]
Static Binaries are now available for download. This way you (and others) can test lrzip-next without having to compile. x64 only.

[Download Here](https://peterhyman.com/download/lrzip-next/)

Enjoy.

**Comment by 8465231** [comment]
Thanks a ton! That works great! I will play with that later today or tomorrow. Got to rebuild the rear end of the truck today.

Are the output files backwards compatible with the normal lrzip?

I actually spent several more hours trying to get it to compile yesterday. It might be an ubuntu issue as I reinstalled 3 times but never saw any type of development option, they only seem to have a "desktop" and "server" release.

I would love to read the manual but in all my google fu (which is quite weak since I am not sure what keywords to use) I could not find anything useful.

My guess is I need to use another distro entirely but ubuntu / Mint are the only linux distros I have ever used (and those for ~6 months of mild use).

I see arch mentioned a lot when looking up development distros, would that be a better option for a base distro? Ideally I want to make a VM with every possible dev tool installed so I can compile any code I want without having to worry about missing packages or dependencies. 

I am actually shocked with the thousands of distros floating around there is not one already made for this purpose.

P.S. Sorry for all this hassle. I really have been doing my best to search for answers on my own (easily 5+ hours on google alone yesterday). Thanks for the pre-compiled binary!

**Reply by pete4abw** [reply]
You can can should create a usable symlink for the static binaries of `lrzip-next`. Something like this:

`ln -s lrzip-next-static-0.7.42 lrzn`

Or whatever directories you are using for it. This, so you don't have to type the whole filename each time.

In answer to your questions.
1. Try and follow these instructions: https://tecadmin.net/install-development-tools-on-ubuntu/
2. No. lrzip-next compressed files are NOT backward compatible to version 0.6 and will not decompress with it. Byte 16 in the magic header is used for coding filtering used. lrzip 0.6x does not have filtering. The lzma bytes are starting at Byte 17 in lrzip-next and byte 16 in lrzip.
3. lrzip 0.6x files ARE readable by lrzip-next.
4. In case you're security conscious, the **asc** file is a gpg signature which ensures byte for byte accuracy for the static binary at the time it was created. A sophisticated checksum if you will.
5. Trying to learn with google is counterproductive. If you don't ask the right question, you won't get the right answer. It took me 30 seconds to get your answer. Get a book. Read. Learn.

**Comment by 8465231** [comment]
Setting up a shortcut for it was actually on my list to figure out, thanks for that.

I actually found that link and followed it yesterday but still had errors (although it was what allowed autogen to complete at all, just with errors). Added a whole bunch more dev packages over the course of the day as well but the errors continued. Some people had success changing the autogen.sh from /bin/sh to /bin/bash but this didn't help me.

Yep, well aware google is only as good as what you put in. I don't even know where to find a book on the subject though, I find super technical jargon from time to time but it goes right over my head as I don't understand most of the terms. 

Linux stuff tends to go from a basic overview with no usable information to a 103 course on the inner workings with little in between.

You have been a big help, thank you.
